Add main

fenrir94 opened the pull request:
Make main.cpp

fenrir94 committed:
Add main

Make main.cpp

fenrir94 committed:
Modify main

Copy barebones.cpp code

fenrir94 committed:
Add pongGame

Make pongGame.h
Make pongGame.cpp

fenrir94 committed:
Print Press AnyKey

Print Press AnyKey text in Screen

fenrir94 committed:
Start Game

- Declare static variable bool gamePlay in pongGame.h
- Initialize gamePlay = false in main.cpp
- Declare function void checkGamePlay in pongGame.h
- Implement checkGameplay in pongGame.cpp

fenrir94 committed:
checkGamePlay

- change function checkGamePlay to static function

fenrir94 committed:
pongGame Start

- Implement fucntion init_PongGame() in pongGame.cpp
- Inplement function uqdate_PongGame() in pongGame.cpp

fenrir94 committed:
Restart Game

- Declare function bool restartGamePlay
- Implement function bool restartGamePlay

fenrir94 committed:
Print Play Time

- Declare and Implement function void  convertTimetoText
- Declare and Implement function void  printTime

fenrir94 committed:
Change pFont

- Make viewport.h/.cpp
- Add static variable s8 pFont
- Modify pFont to viewport::pFont

minjeong03 commented on Alpha/viewport.cpp:
```diff
@@ -0,0 +1,6 @@
+#include "viewport.h"
+
+void viewport::init_viewport()
```
함수 naming convention이 일관적이지 않은데 이유가 있나요?  (e.g funcName, func_name)

fenrir94 replied:
수정하겠습니다.

minjeong03 commented on Alpha/main.cpp:
```diff
@@ -0,0 +1,77 @@
+
+viewport vp;
+bool pongGame::gamePlay = false;
+s8 viewport::pFont;
```
`pFont` 변수는 `viewport` 클래스의 정적 멤버(static member variable)인데,
보통 이런 정적 멤버 변수는 그 클래스의 `.cpp` 파일 — 이 경우라면 `viewport.cpp` — 안에서 정의해주는 게 일반적인 방식인데,

이 변수를 `main.cpp`에서 정의할 특별한 이유가 있나요.

`pongGame::gamePlay` 도 마찬가지

minjeong03 commented on Alpha/pongGame.h:
```diff
@@ -0,0 +1,25 @@
+class pongGame
+{
+public:
+	static bool gamePlay;
```
이 변수 왜 static 변수에요?

fenrir94 replied:
pongGame은 pongGame 게임 1판을 위해 생성되고 해당 게임을 관리하기 위한 클래스이고 gamePlay는 플레이어의 게임 시작을 위한 입력(Spacer Bar)에 따라 pongGame이 동작 중인지 체크하기 위한 bool 변수입니다. 
gamePlay를 pongGame의 private 멤버 변수로 선언하고 pongGame을 전역으로 선언하는 방법도 있지만 게임 1판을 위한 pongGame 객체가 gamePlay 변수를 private로 가지는 것은 옳지 않다고 생각하여 static 전역 변수로 설정했습니다.

minjeong03 commented on Alpha/main.cpp:
```diff
@@ -0,0 +1,77 @@
+		if (pongGame::gamePlay) {
+			pongGame pG;
+			pG.init_PongGame();
+			pG.update_PongGame();
```
main game loop가 update_PongGame() 함수에도 있던데, 자세한 리뷰는 update_PongGame() 함수에 달아놓겠슴다.

minjeong03 commented on Alpha/pongGame.cpp:
```diff
@@ -0,0 +1,90 @@
+void pongGame::update_PongGame()
+{
+    while (pongGame::gamePlay) {
+        AESysFrameStart();
```
update_PongGame()은 한 프레임 단위의 상태 update하는게 올바른 동작일 거 같슴다.

문제1. main game loop가 다수가 됨. 관리 용이하지 않음. (다른 game state가 또 생기면 그 update에서 또 main loop 구현하실 건가요?)
문제2. AESysFrameEnd();하고 나온 뒤 main.cpp 의 while문에서 또 AESysFrameEnd(); 한번 더 하게 됨.

fenrir94 replied:
main의 loop는 게임 전체의 loop이자 메인 대기 화면의  update 기능으로 작성되었습니다. 
game state(main, pongGame)에 따라서 game state 변경시 init함수를 1번 호출하고 update함수가 반복 호출하도록 수정하겠습니다.

fenrir94 committed:
Fix Function Names

- Uniform Function Names in pongGame

fenrir94 committed:
AGameState

Add abstract class AGameState

fenrir94 committed:
pongManager

- Make pongManager to control entire game system

fenrir94 committed:
Remove viewport

- Remove viewport.h and viewport.cpp

fenrir94 committed:
Add utilityKey

- Make utilityKey.h and utility.c
- make function bool isStart_PongGame()
- make function bool isRestart_PongGame()

fenrir94 committed:
Fix AGameState

- Fix funcitons names
- Change ~AGameState to virtual
- Change static gameState to pointer

fenrir94 committed:
Fix static gameState

- Fix pongManager::gameState to pointer
- Remove gameState in AGameState.cpp

fenrir94 committed:
Modify pongGame

- Remove previous functions
- Add override functions of AGameState

fenrir94 committed:
Add mainMenu

- Add mainMenu.h and mainMenu.cpp
- mainMenu is a concrete class of AGameState
- Implement member functions

fenrir94 committed:
Modify main

- Modify function wWinMain in main.cpp to apply state pattern with pongManager

fenrir94 committed:
Modify pongManager

- Modify function call structure to fix screen print defect

fenrir94 committed:
pongGameState

- Declare extern pongGameState

fenrir94 committed:
Modify mainMenu

- Modify Update() to change gameState

fenrir94 committed:
Modify pongGame

- Modfiy Function call structure

fenrir94 committed:
Modify update

- Add if condition to restart pongGame

fenrir94 committed:
modify update

- Call exit() when gameState is changed

minjeong03 commented on Alpha/mainMenu.cpp:
```diff
@@ -22,7 +23,9 @@ void mainMenu::initialize()
 void mainMenu::update()
 {
-
+	if (isStart_PongGame()) {
```
이거는 MainMenu에 특정한 기능인 것 같은데 그럼, main menu.cpp에 있어야 할 것 같음.

fenrir94 replied:
수정했습니다.

minjeong03 commented on Alpha/pongGame.cpp:
```diff
@@ -25,35 +27,21 @@ void pongGame::initialize()
-		pongGame::gamePlay = false;
-		break;
+		pongManager::changeGameState_PongManager(&pongGameState);
+		exit();
```
offline으로 논의 함. 다음 commit에서 다시 얘기하도록 함.

minjeong03 commented on Alpha/pongManager.cpp:
```diff
@@ -48,10 +49,14 @@ void pongManager::update_PongManager()
 {
 	
 	while (1) {
```
while(1) 무한루프라면, 

SysExit() <-- 요건 어디서 불리게 되는 걸까요?

minjeong03 replied:
break statement를 못봤음. srry

fenrir94 replied:
pongManager에 멤버 변수 isGameRunning를 추가하고 while 루프 조건 1을 대체하여 수정했습니다.

fenrir94 committed:
Fix key event functions

- Move function isStart_PongGame to class mainMenu
- Move function isRestart_PongGame to class pongGame

fenrir94 committed:
Fix loop condition

- Change 1 to a bool variable, isGameRunning
